packages/opentelemetry-plugin-mongodb/src/mongodb.ts

```diff
@@ -0,0 +1,208 @@
+/* tslint:disable:deprecation */
+
+import { BasePlugin } from '@opentelemetry/core';
+import { Span, SpanKind, CanonicalCode } from '@opentelemetry/types';
+import * as mongodb from 'mongodb';
+import * as shimmer from 'shimmer';
+
+type Func<T> = (...args: unknown[]) => T;
+type MongoDB = typeof mongodb;
+interface MongoInternalCommand {
+  findandmodify: boolean;
+  createIndexes: boolean;
+  count: boolean;
+  ismaster: boolean;
+}
+
+/** MongoDB instrumentation plugin for OpenTelemetry */
+export class MongoDBPlugin extends BasePlugin<MongoDB> {
+  private readonly _SERVER_METHODS = ['insert', 'update', 'remove', 'command'];
+  private readonly _CURSOR_METHODS = ['_next', 'next'];
+
+  protected readonly _supportedVersions = ['>=2 <3'];
+
+  constructor(readonly moduleName: string) {
+    super();
+  }
+
+  /**
+   * Patches MongoDB operations.
+   */
+  protected patch() {
+    this._logger.debug('Patched MongoDB');
+
+    if (this._moduleExports.Server) {
+      for (const fn of this._SERVER_METHODS) {
+        this._logger.debug(`patching mongodb-core.Server.prototype.${fn}`);
+        shimmer.wrap(
+          this._moduleExports.Server.prototype,
+          // Forced to ignore due to incomplete typings
+          // tslint:disable-next-line:ban-ts-ignore
+          // @ts-ignore
+          fn,
+          this._getPatchCommand(fn)
+        );
+      }
+    }
+
+    if (this._moduleExports.Cursor) {
+      this._logger.debug(
+        'patching mongodb-core.Cursor.prototype functions:',
+        this._CURSOR_METHODS
+      );
+      shimmer.massWrap(
+        [this._moduleExports.Cursor.prototype],
+        this._CURSOR_METHODS as never[],
+        // tslint:disable-next-line:no-any
+        this._getPatchCursor() as any
+      );
+    }
+
+    return this._moduleExports;
+  }
+
+  /** Unpatches all MongoDB patched functions. */
+  unpatch(): void {
+    shimmer.massUnwrap([this._moduleExports.Server.prototype], this
+      ._SERVER_METHODS as never[]);
+    shimmer.massUnwrap([this._moduleExports.Cursor.prototype], this
+      ._CURSOR_METHODS as never[]);
+  }
+
+  /** Creates spans for Command operations */
+  private _getPatchCommand(operationName: string) {
+    const plugin = this;
+    return (original: Func<mongodb.Server>) => {
+      return function patchedServerCommand(
+        this: mongodb.Server,
+        ns: string,
+        command: MongoInternalCommand,
+        options: {} | Function,
+        callback: Func<unknown>
+      ): mongodb.Server {
+        const currentSpan = plugin._tracer.getCurrentSpan();
+        // @ts-ignore
+        const resultHandler =
+          typeof options === 'function' ? options : callback;
+        if (
+          currentSpan === null ||
+          typeof resultHandler !== 'function' ||
+          typeof command !== 'object'
+        ) {
+          return original.apply(this, (arguments as unknown) as unknown[]);
+        }
+        let type: string;
```

Comment:
Could you please split this logic into smaller function? Maybe, `const type = getType(command);`